src/utransport.rs

```diff
@@ -230,38 +240,47 @@ pub trait UTransport: Send + Sync {
     /// #     }
     /// # }
     /// #
-    /// # let my_transport = MyTransport::new();
-    /// # let my_uuri = UUri::default();
+    /// # let mut my_transport = MyTransport::new();
+    /// # let source_uuri = UUri::default();
     ///
     /// // hang onto this listener...
-    /// let my_listener: Arc<dyn UListener> = Arc::new(MyListener::new());
-    /// // ...send a clone through when registering...
-    /// let register_result = my_transport.register_listener(my_uuri.clone(), my_listener.clone());
+    /// let my_listener = Arc::new(MyListener::new());
+    /// // ...use a clone for registering...
+    /// my_transport.register_listener(
+    ///     Some(&source_uuri),
+    ///     None,
+    ///     my_listener.clone());
     /// // ...and use the original we hung onto when unregistering
-    /// let unregister_result = my_transport.unregister_listener(my_uuri, my_listener);
+    /// my_transport.unregister_listener(
+    ///     Some(&source_uuri),
+    ///     None,
+    ///     my_listener);
     /// ```
     async fn register_listener(
-        &self,
-        topic: UUri,
+        &mut self,
```

Comment:
Why is an exclusive reference needed instead of a shared reference in the new formulation?

Reply:
I think this is because there is no Mutex here, so needs to be mutable to modify the hash table.
https://github.com/eclipse-uprotocol/up-rust/pull/92/files#diff-2b11318fb2673b25403b992883ee6ac91316e6f11f3c32c1b352caf4e07c79beR358

Reply:
yes, that is the reason. Since un-/registering will (generally speaking) require some modification of the transport's state, I do not think that this is an issue but actually provides more options for implementation (as we see in the unit tests). WDYT?

Reply:
I'm fine with that. We can use `Arc` + `Mutex` to `UpClientXXX` if need to pass to other threads.

Reply:
Ok, it turns out that using `&mut self` here is not such a great idea because it requires client code to get a grip on a mutable `Arc<dyn UTransport>` which is not as easy as it sounds. The [Arc::get_mut documentation](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.get_mut) tells us why. I will change this back to just `&self` in order to make implementation of client code easier, even if this means some additional effort for implementors of UTransport.